wasm/src/lib.rs

```diff
@@ -1,33 +1,133 @@
+            rustpython_vm::obj::objfloat::make_float(vm, &int)
+                .unwrap()
+                .into()
+        }
+        PyObjectKind::Float { ref value } => JsValue::from_f64(*value),
+        PyObjectKind::Bytes { ref value } => {
+            let arr = js_sys::Uint8Array::new(&JsValue::from(value.len() as u32));
+            for (i, byte) in value.iter().enumerate() {
+                console::log_1(&JsValue::from(i as u32));
+                js_sys::Reflect::set(&arr, &JsValue::from(i as u32), &JsValue::from(*byte))
+                    .unwrap();
+            }
+            arr.into()
+        }
+        PyObjectKind::Sequence { ref elements } => {
+            let arr = js_sys::Array::new();
+            for val in elements {
+                arr.push(&py_to_js(vm, val));
+            }
+            arr.into()
+        }
+        PyObjectKind::Dict { ref elements } => {
+            let obj = js_sys::Object::new();
+            for (key, (_, val)) in elements {
+                js_sys::Reflect::set(&obj, &key.into(), &py_to_js(vm, val))
+                    .expect("couldn't set property of object");
+            }
+            obj.into()
+        }
+        PyObjectKind::None => JsValue::UNDEFINED,
+        _ => JsValue::UNDEFINED,
+    }
+}
+
+fn eval(vm: &mut VirtualMachine, source: &str) -> PyResult {
+    // HACK: if the code doesn't end with newline it crashes.
+    let mut source = source.to_string();
+    if !source.ends_with('\n') {
+        source.push('\n');
+    }
+
+    let code_obj = compile::compile(vm, &source, compile::Mode::Exec, None)?;
+
+    let builtins = vm.get_builtin_scope();
+    let vars = vm.context().new_scope(Some(builtins));
+
+    vm.run_code_obj(code_obj, vars)
+}
+
+#[wasm_bindgen]
+pub fn eval_py(source: &str) -> Result<JsValue, JsValue> {
+    let mut vm = VirtualMachine::new();
+
+    vm.ctx.set_attr(
+        &vm.builtins,
+        "print",
+        vm.context().new_rustfunc(wasm_builtins::builtin_log),
+    );
+
+    eval(&mut vm, source)
+        .map(|value| py_to_js(&mut vm, &value))
+        .map_err(|err| py_str_err(&mut vm, &err).into())
+}
+
 #[wasm_bindgen]
-pub fn run_code(source: &str) -> () {
+pub fn run_from_textbox(source: &str) -> Result<JsValue, JsValue> {
     //add hash in here
     console::log_1(&"Running RustPython".into());
     console::log_1(&"Running code:".into());
     console::log_1(&source.to_string().into());
 
     let mut vm = VirtualMachine::new();
-    // We are monkey-patching the builtin print to use console.log
-    // TODO: moneky-patch sys.stdout instead, after print actually uses sys.stdout
-    vm.builtins.set_attr("print", vm.context().new_rustfunc(wasm_builtins::builtin_print));
 
-    let code_obj = compile::compile(&mut vm, &source.to_string(), compile::Mode::Exec, None);
+    // We are monkey-patching the builtin print to use console.log
+    // TODO: monkey-patch sys.stdout instead, after print actually uses sys.stdout
+    vm.ctx.set_attr(
```

Comment:
Oh, thanks! I did not realize I broke this..

Reply:
🤗